### PR TITLE
[ci] Add coq-http

### LINKED ISSUE #16093: Add coq-http to CI
We should add https://github.com/liyishuai/coq-http to the CI.

As recommended here: https://github.com/coq/coq/pull/15333#discussion_r884178848. Although I am not sure we would be able to catch the issue with extraction on an older version of OCaml.

Which versions of OCaml are we using in the CI @coq/ci-maintainers?

### PR DESCRIPTION
**Pull request by liyishuai**
Closes #16093

- [x] Added / updated **test-suite**.
- [x] #17689

### CODE REVIEW

**Comment by liyishuai** [comment]
Expect [compilation failure](https://app.circleci.com/pipelines/github/QuickChick/QuickChick/611/workflows/c7b8385d-1906-48c1-9d25-5c0ae521f491/jobs/4298?invite=true#step-116-68) with Coq dev:
> COQC app/Encode.v
File "./app/Encode.v", line 6, characters 0-40:
Error:
Dynlink error: execution of module initializers in the shared library failed: "Anomaly: bad vernac extend: Derive, 0"

Need help in fixing coq-http.

**Comment by SkySkimmer** [comment]
Tried locally, I get
~~~
File "./app/AppMessage.v", line 5, characters 0-39:
Error: Cannot find a physical path bound to logical path
Parser with prefix Parsec.
File "./theories/Printer.v", line 7, characters 0-37:
Error: Cannot find a physical path bound to logical path
Ceres with prefix Ceres.
~~~

You can try locally with 
~~~bash
git clone (your branch of coq)
cd coq
make world
make -j (however much your machine can handle) ci-http
~~~

**Comment by liyishuai** [comment]
Should I add Parsec and Ceres to the Coq CI as well? or are there other ways to add OPAM dependencies?

**Comment by SkySkimmer** [comment]
All dependencies need to be in CI otherwise it won't work

**Comment by liyishuai** [comment]
Resuming this thread after paper submission deadline…

**Comment by Alizter** [comment]
@liyishuai ping

**Comment by liyishuai** [comment]
Still WIP. Will press "Ready for review" once it really is.

**Comment by liyishuai** [comment]
Too many packages to add. Let's do #17689 first.

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by Alizter** [comment]
@coqbot run full ci

**Comment by liyishuai** [comment]
What fix should I perform?
```
make[1]: Leaving directory '/builds/coq/coq/_build_ci/ext_lib'
Aggregating timing log...
No timing data
Uploading artifacts for failed job
Uploading artifacts...
_build_ci: found 9225 matching files and directories 
Uploading artifacts as "archive" to coordinator... 201 Created  id=4468263521 responseStatus=201 Created token=64_yyMHi
Cleaning up project directory and file based variables
ERROR: Job failed: exit code 1
```

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: Makefile.ci [hunk 1]
```diff
@@ -133,6 +135,12 @@ ci-geocoq: ci-mathcomp
 ci-simple_io: ci-ext_lib
 ci-quickchick: ci-ext_lib ci-simple_io ci-mathcomp
 ci-itree: ci-ext_lib ci-paco
+ci-itree_io: ci-itree ci-simple_io
+
+ci-parsec: ci-ext_lib ci-ceres
+ci-json: ci-parsec ci-menhir
+ci-async_test: ci-itree_io ci-json ci-quickchick
+ci-http: ci-async_test
```
Comment (liyishuai) [inline Makefile.ci:143]: > File "./app/AppMessage.v", line 5, characters 0-39:
Error: Cannot find a physical path bound to logical path
Parser with prefix Parsec.

I'm getting confused. How to make `ci-http` aware of Parsec? I thought `Makefile.ci` did the jobs.

Comment (SkySkimmer) [reply Makefile.ci:143]: ci-parsec isn't in the CI_TARGETS list in the makefile (and it's not detected because there's `ci-parsec: ci-ext_lib ci-ceres`)

**Comment by SkySkimmer** [comment]
Are we sure we want this in CI? All those dependencies seem like it could get annoying if we need to overlay.

**Comment by Alizter** [comment]
The reason we wanted to add it in the first place is to catch breakages we caused, especially in extraction.

**Comment by Zimmi48** [comment]
@coqbot run full CI

@liyishuai do you want to be added to the `@coq/contributors` team to be able to call these commands yourself directly? (See https://github.com/coq/coq/blob/master/CONTRIBUTING.md#helping-triage-existing-issues for what this means and https://github.com/coq/coq/blob/master/CONTRIBUTING.md#understanding-automatic-feedback for the coqbot commands).

**Comment by liyishuai** [comment]
> @liyishuai do you want to be added to the `@coq/contributors` team to be able to call these commands yourself directly?

Yes I'd like to take that responsibility.

**Comment by Zimmi48** [comment]
> > @liyishuai do you want to be added to the `@coq/contributors` team to be able to call these commands yourself directly?
> 
> Yes I'd like to take that responsibility.

Very sorry @liyishuai, I had missed your message. You have now been invited to join the team.

**Comment by Zimmi48** [comment]
Note that rebasing is preferred when resolving conflicts.

**Comment by liyishuai** [comment]
@coqbot run full CI

**Comment by liyishuai** [comment]
Are there pending tasks before reviewing this PR?

**Comment by liyishuai** [comment]
@coqbot run full CI

**Comment by SkySkimmer** [comment]
It seems to work, I just don't know how useful it is to add this to CI.

**Comment by liyishuai** [comment]
As mentioned in #16093, coq-http revealed an extraction error in https://github.com/coq/coq/pull/15333#discussion_r884178848.
Compiling the HTTP tester provides nontrivial coverage of the Coq extraction mechanism (92 modules).

**Comment by Zimmi48** [comment]
Given that this only takes 10 minutes to run, this seems like a sufficient justification for inclusion to me.

**Comment by SkySkimmer** [comment]
Since quickchick now has dependents running the tests should be done in a separate job (see mathcomp_test).

Once that's done I think we can merge.

**Comment by liyishuai** [comment]
@coqbot run full CI

**Comment by liyishuai** [comment]
@coqbot run full CI

**Comment by liyishuai** [comment]
@coqbot run full CI

**Review by SkySkimmer — APPROVED** [review]
Seems to work

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**